types/redux-pack/index.d.ts

```diff
@@ -1,52 +1,77 @@
+    readonly START: 'start';
+    readonly SUCCESS: 'success';
+    readonly FAILURE: 'failure';
 };
-export type LIFECYCLEValues = 'start'| 'succes'| 'failure';
+
+export type LIFECYCLEValues = 'start' | 'succes' | 'failure';
 
 export const middleware: Middleware;
 
-export interface Handlers<S> {
-    start?: Reducer<S>;
-    finish?: Reducer<S>;
-    failure?: Reducer<S>;
-    success?: Reducer<S>;
-    always?: Reducer<S>;
+// MetaPayload differs from ActionMeta in that it is the object that the reducers
+// receive, instead of what is dispatched
+export type MetaPayload<M> = M & {
+    ['redux-pack/LIFECYCLE']?: LIFECYCLEValues;
+    ['redux-pack/TRANSACTION']?: string;
+};
+
+// Incomplete typing
+export type PackActionPayload<Payload, M> = ReduxAction & {
+    payload: Payload;
+    meta: MetaPayload<M>;
+};
+
+export type handlerReducer<S, A> = (state: S, action: A) => S;
+export interface Handlers<S, TSuccessPayload, TErrorPayload, TStartPayload, TMetaPayload> {
+    start?: handlerReducer<S, PackActionPayload<TStartPayload, TMetaPayload>>;
+    finish?: handlerReducer<S, ReduxAction>;
+    failure?: handlerReducer<S, PackActionPayload<TErrorPayload, TMetaPayload>>;
+    success?: handlerReducer<S, PackActionPayload<TSuccessPayload, TMetaPayload>>;
+    always?: handlerReducer<S, ReduxAction>;
 }
+
 export type GetState<S> = () => S;
-export interface ActionMeta<TState = {}, TSuccessPayload = {}, TErrorPayload = {}, TStartPayload = {}> {
+export interface ActionMeta<TFullState = {}, TSuccessPayload = {}, TErrorPayload = {}, TStartPayload = {}> {
     startPayload?: TStartPayload;
-    onStart?(payload: TStartPayload, getState: GetState<TState>): void;
-    onFinish?(resolved: boolean, getState: GetState<TState>): void;
-    onSuccess?(response: TSuccessPayload, getState: GetState<TState>): void;
-    onFailure?(error: TErrorPayload, getState: GetState<TState>): void;
-    ['redux-pack/LIFECYCLE']?: keyof LIFECYCLEValues;
+    onStart?(payload: TStartPayload, getState: GetState<TFullState>): void;
+    onFinish?(resolved: boolean, getState: GetState<TFullState>): void;
+    onSuccess?(response: TSuccessPayload, getState: GetState<TFullState>): void;
+    onFailure?(error: TErrorPayload, getState: GetState<TFullState>): void;
+    ['redux-pack/LIFECYCLE']?: LIFECYCLEValues;
     ['redux-pack/TRANSACTION']?: string;
 }
-export interface Action<TState = {}, TSuccessPayload = {}, TErrorPayload = {}, TStartPayload = {}> extends ReduxAction {
+
+export interface PackError { error: boolean; payload: any; }
+export interface Action<TFullState = {}, TSuccessPayload = {}, TErrorPayload = PackError, TStartPayload = {}, TMetaPayload = {}> extends ReduxAction {
     promise?: Promise<TSuccessPayload>;
     payload?: TSuccessPayload | TErrorPayload | TStartPayload;
-    meta?: ActionMeta<TState, TSuccessPayload, TErrorPayload, TStartPayload>;
+    meta?: ActionMeta<TFullState, TSuccessPayload, TErrorPayload, TStartPayload> & TMetaPayload;
```

Comment:
@tansongyang the addition of the `TMetaPayload` argument to the Action generic and the intersection typing of `ActionMeta` and `TMetaPayload` on this line (along with @quicksnap's new `PackActionPayload` and `MetaPayload` typings) seem to correctly solve the original use case of passing custom metadata keys through to the reducer (without adding the key signature to the `ActionMeta` type def as I originally did. This seems like the safer and better solution.